tools/celtest/test_runner.go

```diff
@@ -0,0 +1,708 @@
+type TestResult struct {
+	Success bool
+	Wanted  string
+	Error   error
+}
+
+// NewTestRunner creates a Test Runner with the provided options.
+// The options can be used to:
+// - configure the Compiler used for parsing and compiling the input expressions
+// - configure the Test Runner used for parsing and executing the tests
+func NewTestRunner(opts ...TestRunnerOption) (*TestRunner, error) {
+	tr := &TestRunner{}
+	var err error
+	for _, opt := range opts {
+		tr, err = opt(tr)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return tr, nil
+}
+
+// AddFileDescriptorSet creates a Test Runner Option which adds a file descriptor set to the test
+// runner. The file descriptor set is used to register proto messages in the global proto registry.
+func AddFileDescriptorSet(path string) TestRunnerOption {
+	return func(tr *TestRunner) (*TestRunner, error) {
+		if path != "" {
+			tr.FileDescriptorSetPath = path
+		}
+		return tr, nil
+	}
+}
+
+func registerMessages(path string) error {
+	if path == "" {
+		return nil
+	}
+	fds, err := fileDescriptorSet(path)
+	if err != nil {
+		return err
+	}
+	for _, file := range fds.GetFile() {
+		reflectFD, err := protodesc.NewFile(file, protoregistry.GlobalFiles)
+		if err != nil {
+			return fmt.Errorf("protodesc.NewFile(%q) failed: %v", file.GetName(), err)
+		}
+		if _, err := protoregistry.GlobalFiles.FindFileByPath(reflectFD.Path()); err == nil {
+			continue
+		}
+		err = protoregistry.GlobalFiles.RegisterFile(reflectFD)
+		if err != nil {
+			return fmt.Errorf("protoregistry.GlobalFiles.RegisterFile() failed: %v", err)
+		}
+		for i := 0; i < reflectFD.Messages().Len(); i++ {
+			msg := reflectFD.Messages().Get(i)
+			msgType := dynamicpb.NewMessageType(msg)
+			err = protoregistry.GlobalTypes.RegisterMessage(msgType)
+			if err != nil && !strings.Contains(err.Error(), "already registered") {
+				return fmt.Errorf("protoregistry.GlobalTypes.RegisterMessage(%q) failed: %v", msgType, err)
+			}
+		}
+	}
+	return nil
+}
+
+func fileDescriptorSet(path string) (*descpb.FileDescriptorSet, error) {
+	bytes, err := os.ReadFile(path)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read file descriptor set file %q: %v", fileDescriptorSetPath, err)
+	}
+	fds := &descpb.FileDescriptorSet{}
+	if err := proto.Unmarshal(bytes, fds); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal file descriptor set file %q: %v", fileDescriptorSetPath, err)
+	}
+	return fds, nil
+}
+
+// Programs creates a list of CEL programs from the input expressions configured in the test runner
+// using the provided program options.
+func (tr *TestRunner) Programs(t *testing.T, opts ...cel.ProgramOption) ([]cel.Program, error) {
+	t.Helper()
+	if tr.Compiler == nil {
+		return nil, fmt.Errorf("compiler is not set")
+	}
+	e, err := tr.CreateEnv()
+	if err != nil {
+		return nil, err
+	}
+	var programs []cel.Program
+	for _, expr := range tr.Expressions {
+		ast, _, err := expr.CreateAST(tr.Compiler)
```

Comment:
Can we add a TODO to propagate the metadata from CreateAST back with the Program instances? Possibly as another struct which combines this information?